### PR TITLE
Add resource name decryption for Crypto.

### PR DESCRIPTION
**Pull request by XODE0**
Decrypt resource name for constants decryption.

### CODE REVIEW

File: de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs [hunk 1]
```diff
@@ -46,5 +47,74 @@ static class CoUtils {
 			}
 			return null;
 		}
+
+        public static string XorCipher(string text, int key) {
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs]: Tabs should be used.

File: de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs [hunk 2]
```diff
@@ -46,5 +47,74 @@ static class CoUtils {
...
+            return string.Intern(new string(array));
+        }
+
+        public static string DecryptResourceName(ModuleDefMD module ,MethodDef method) {
+            string ResourceName = "";
+            MethodDef cctor = method, OrginalResMethod = null;
+            //retrive key and encrypted resource name 
+            int key = 0;
+            var ILs = cctor.Body.Instructions;
+            for (int i = 0; i < ILs.Count; i++) {
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs]: Should be i+2 < ILs.Count

File: de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs [hunk 3]
```diff
@@ -46,5 +47,74 @@ static class CoUtils {
...
+            int num = resourceName.Length;
+            char[] array = resourceName.ToCharArray();
+            while (--num >= 0) {
+                array[num] = (char)((int)array[num] ^ ((int)coddedBytes[key & 15] | key));
+            }
+            return string.Intern(new string(array));
+        }
+
+        public static string DecryptResourceName(ModuleDefMD module ,MethodDef method) {
+            string ResourceName = "";
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs]: First char of all locals should be lower case

File: de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs [hunk 4]
```diff
@@ -46,5 +47,74 @@ static class CoUtils {
...
+            return new string(array);
+        }
+
+        public static string DecryptResourceName(string resourceName, int key, byte[] coddedBytes) {
+            int num = resourceName.Length;
+            char[] array = resourceName.ToCharArray();
+            while (--num >= 0) {
+                array[num] = (char)((int)array[num] ^ ((int)coddedBytes[key & 15] | key));
+            }
+            return string.Intern(new string(array));
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/CryptoObfuscator/CoUtils.cs]: Why do you call string.Intern()?

File: de4dot.code/deobfuscators/CryptoObfuscator/ConstantsDecrypter.cs [hunk 5]
```diff
@@ -113,7 +113,16 @@ class ConstantsDecrypter {
...
-			encryptedResource = CoUtils.GetResource(module, DotNetUtils.GetCodeStrings(decrypterType.FindStaticConstructor()));
+            MethodDef cctor = decrypterType.FindStaticConstructor();
+            encryptedResource = CoUtils.GetResource(module, DotNetUtils.GetCodeStrings(cctor));
+
+            //if the return value is null, it is possible that resource name is encrypted
+            if (encryptedResource == null) {
+                    var Resources = new string[] { CoUtils.DecryptResourceName(module,cctor) };
+                    encryptedResource = CoUtils.GetResource(module, Resources);
+                } else
+                    Logger.v("Can't get decryption key for resource name");
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/CryptoObfuscator/ConstantsDecrypter.cs]: Remove this message.

File: de4dot.code/deobfuscators/CryptoObfuscator/ConstantsDecrypter.cs [hunk 6]
```diff
@@ -113,7 +113,16 @@ class ConstantsDecrypter {
...
 
-			encryptedResource = CoUtils.GetResource(module, DotNetUtils.GetCodeStrings(decrypterType.FindStaticConstructor()));
+            MethodDef cctor = decrypterType.FindStaticConstructor();
+            encryptedResource = CoUtils.GetResource(module, DotNetUtils.GetCodeStrings(cctor));
+
+            //if the return value is null, it is possible that resource name is encrypted
+            if (encryptedResource == null) {
+                    var Resources = new string[] { CoUtils.DecryptResourceName(module,cctor) };
+                    encryptedResource = CoUtils.GetResource(module, Resources);
+                } else
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/CryptoObfuscator/ConstantsDecrypter.cs]: else should always be on a new line

File: de4dot.code/deobfuscators/CryptoObfuscator/StringDecrypter.cs [hunk 7]
```diff
@@ -87,6 +88,12 @@ class StringDecrypter {
 					return Encoding.UTF8.GetString(Convert.FromBase64String(s));
 				}
 				catch {
+                    string s2 = CoUtils.DecryptResourceName(module, cctor);
+                    try {
+                         return Encoding.UTF8.GetString(Convert.FromBase64String(s2));
+                    } catch {
```
Comment (0xd4d) [inline de4dot.code/deobfuscators/CryptoObfuscator/StringDecrypter.cs]: catch, just like else, should be on a new line. The empty line below is not needed.

**Comment by 0xd4d** [comment]
It passed my tests, but you need to fix your commits, eg. you don't follow the de4dot coding style. Fix, and merge the changes

```
git reset 4c684bb67ec4116343e6a62b424cda2e21f53ea7
# ... fix code
git add .
git commit -m "Your message"
git push --force
```

**Comment by XODE0** [comment]
Most of things are fixed now except for tabs, I can't commit after i select 4(selecting the size 8 won't make any difference) , Do i need to fix it manually?

**Comment by 0xd4d** [comment]
Set tab size to 4 in VS settings (Tools, Options, Text Editor, C#, Tabs), then select the lines and pick Edit, Advanced, Tabify Selected Lines, then redo the commands in my first commit again.

**Comment by XODE0** [comment]
Thank you for your help, everything should be fine now.

**Comment by mathiasbynens** [comment]
@0xd4d Have you heard of EditorConfig? There’s [a plugin available for VS](https://github.com/editorconfig/editorconfig-visualstudio). I’d be happy to submit a pull request with an `.editorconfig` file so that contributors can follow the existing style more easily.

**Comment by 0xd4d** [comment]
It's an optional VS plugin, I don't want to force people into installing a VS plugin just to send a pull request.

**Comment by mathiasbynens** [comment]
Sure. You wouldn’t force it though — it would just be helpful for those who have it installed anyway (no more manual setting switching for each project).

**Pull request merged**